### PR TITLE
setup docker environment

### PR DESCRIPTION
**Pull request by dboyliao**
- Setup docker environment for development
- Add README for instructions of building docker environment

### CODE REVIEW

**Comment by dboyliao** [comment]
FYI, I'm able to build `modmesh` and import it in the docker container.

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[submodule "thirdparty/pybind11"]
```
Comment (yungyuc) [inline .gitmodules]: Dependency should be installed outside the project, except some very special code that will be considered an integrated part in the source form, e.g., [metal-cpp](https://github.com/solvcon/modmesh/blob/master/thirdparty/metal-cpp.sh).

One way to build the dependency is to use https://github.com/solvcon/devenv .

File: Makefile [hunk 2]
```diff
@@ -112,8 +112,15 @@ cmake: $(BUILD_PATH)/Makefile
 .PHONY: xcode
 xcode: $(BUILD_PATH)_xcode/Makefile
 
+.PHONY: pybind11
```
Comment (yungyuc) [inline Makefile]: pybind11 is a header only library and should not use a target.  Please revert the change in the make file.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# Docker
```
Comment (yungyuc) [inline README.md]: We would like to avoid use of container because it is hard to maintain.  Container is good for deployment but it is cumbersome for source-level debugging.  For example, it's hard to compare the debug and optimized builds using a container.  It also adds complexity and overheads in instrumenting and profiling.

It's probably OK to add optional scripts for container, but we should not bless the use.

File: docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,32 @@
+FROM ubuntu:22.04
```
Comment (yungyuc) [inline docker/Dockerfile]: If adding docker script, it should be put under `contrib/`, not the root, because container is not a blessed way to build it.

**Review by yungyuc — CHANGES_REQUESTED** [review]

File: Makefile (hunk 2 above)
Comment (dboyliao) [reply Makefile]: reverted.

File: .gitmodules (hunk 1 above)
Comment (dboyliao) [reply .gitmodules]: Moving docker files into `/contrib`

File: docker/Dockerfile (hunk 4 above)
Comment (dboyliao) [reply docker/Dockerfile]: Done

**Comment by dboyliao** [comment]
Screen Shot 2022-09-11 at 3 04 55 PM

I run the CI locally and the error is due to missing`pip3` command.
Looks like `pip` is not installed correctly in the CI image

**Comment by yungyuc** [comment]
> Screen Shot 2022-09-11 at 3 04 55 PM
> 
> I run the CI locally and the error is due to missing`pip3` command. Looks like `pip` is not installed correctly in the CI image

Could you please provide a link to the error at your CI run?  Image is not helpful in pin-pointing the error.

The nightly CI is healthy: https://github.com/solvcon/modmesh/actions/runs/3028860237 .  I would guess the error comes from your change.

File: .gitmodules (hunk 1 above)
Comment (yungyuc) [reply .gitmodules]: Perhaps I didn't make it clear.  What I meant was that there should not be sub-module in the repository and devenv is one way we did to avoid it.

File: README.md (hunk 3 above)
Comment (yungyuc) [reply README.md]: Since container is not a blessed way to build the code, it should not be mentioned in README especially when a blessed approach is not included.

That is, at the moment, the project is not ready for a README.  This file should not exist (yet).

File: .gitmodules (hunk 1 above)
Comment (dboyliao) [reply .gitmodules]: I do remove the submodule
Please check the `Files changed` tab for detail.
I'm just telling you I also move all the docker files into `/contrib`.

File: README.md (hunk 3 above)
Comment (dboyliao) [reply README.md]: Ok. I'll also move the README into `/contrib`

**Comment by dboyliao** [comment]
> > Screen Shot 2022-09-11 at 3 04 55 PM
> > I run the CI locally and the error is due to missing`pip3` command. Looks like `pip` is not installed correctly in the CI image
> 
> Could you please provide a link to the error at your CI run? Image is not helpful in pin-pointing the error.
> 
> The nightly CI is healthy: https://github.com/solvcon/modmesh/actions/runs/3028860237 . I would guess the error comes from your change.

I'm using `act` to run the CI locally.
https://github.com/nektos/act

After inspecting the workflow, that's because I use `pyenv` on my machine and it's fighting with it.
My PR should not introduce any CI-breaking change after all.

File: .gitmodules (hunk 1 above)
Comment (yungyuc) [reply .gitmodules]: But the file is still in your changeset: https://github.com/dboyliao/modmesh/commit/730bef36c6e76806bd7c9da9e6385c42bea28f89

File: README.md (hunk 3 above)
Comment (yungyuc) [reply README.md]: In that way, how about adding the information as the comments in the docker file you add?

File: .gitmodules (hunk 1 above)
Comment (dboyliao) [reply .gitmodules]: Oh, I forgot to remove the file.
My bad.
Sorry.

**Review by yungyuc — APPROVED** [review]

**Comment by yungyuc** [comment]
LGTM.  Thanks for the PR.

**Comment by tigercosmos** [comment]
wow love it!!
